ext/rugged/rugged_patch.c

```diff
@@ -332,16 +333,31 @@ static int patch_print_cb(
 	const git_diff_line *line,
 	void *payload)
 {
-	VALUE rb_str = (VALUE)payload;
+	VALUE rb_buffer = (VALUE)payload;
 
 	switch (line->origin) {
 		case GIT_DIFF_LINE_CONTEXT:
 		case GIT_DIFF_LINE_ADDITION:
 		case GIT_DIFF_LINE_DELETION:
-			rb_str_cat(rb_str, &line->origin, 1);
+			rb_ary_push(rb_buffer, rb_str_new(&line->origin, 1));
 	}
 
-	rb_str_cat(rb_str, line->content, line->content_len);
+	rb_ary_push(rb_buffer, rb_str_new(line->content, line->content_len));
+
+	return GIT_OK;
+}
+
+static int patch_print_header_cb(
+	const git_diff_delta *delta,
+	const git_diff_hunk *hunk,
+	const git_diff_line *line,
+	void *payload)
+{
+	VALUE rb_buffer = (VALUE)payload;
+
+	if (line->origin == GIT_DIFF_LINE_FILE_HDR) {
+		rb_ary_push(rb_buffer, rb_str_new(line->content, line->content_len));
+	}
 
 	return GIT_OK;
```

Comment:
What do you think about returning `GIT_EUSER` (or I wonder if `GIT_ITEROVER` would work too?) once we hit the first line after the header?

That should stop the internal print loop from calling this callback once we have what we need. We'll just want to check the return value of `git_patch_print` before passing it to `rugged_exception_check` for this special error code.